update NR difficulty

cooljeanius opened the pull request:
NR is notoriously difficult and annoying, so this PR attempts to soften it a little bit on the easier difficulties (not on HARD, though). These changes were originally made during my 1.14 playthru of NR, and haven't been tested on 1.15/1.16/1.17 yet, although I'm really dreading having to test them again... I'm leaving this marked as a draft because there are rejects files for map files that I haven't managed to wiggle back in yet, because this should really wait for until after ~~1.16~~ ***1.18*** is released, and because I still need to do another pass at editing/testing.

cooljeanius committed:
initial attempt at applying patches from my repo

see https://github.com/cooljeanius/wesnoth_mods/tree/master/campaigns/Northern_Rebirth/diffs
NB there are some rejects files that I'll have to wiggle in

cooljeanius committed:
wiggle in rejects files for scenario files

there are still rejects files for maps though

cooljeanius committed:
re-linearize some numerical progressions

cooljeanius committed:
Update 13a_Showdown.cfg

minor comment movement, for fixing diff display

cooljeanius committed:
Update 13a_Showdown.cfg

still trying to fix diff display

cooljeanius committed:
Update 13a_Showdown.cfg

more updating of commentary for diff display purposes; hopefully this will be the final one...

cooljeanius commented:
These changes originally come from here, btw: https://github.com/cooljeanius/wesnoth_mods/tree/master/campaigns/Northern_Rebirth/diffs

kabachuha commented on data/campaigns/Northern_Rebirth/scenarios/08a_Elvish_Princess.cfg:
```diff
@@ -94,9 +94,13 @@
         controller=human
-        recruit=Thug,Poacher,Spearman,Bowman,Gryphon,Dwarvish Fighter,Dwarvish Thunderer,Dwarvish Ulfserker,Dwarvish Scout,Footpad
+        # Tallin only asked the mages for their gryphons at the end of the previous scenario, so don't add any more of their units here:
+        recruit=Thug,Poacher,Spearman,Bowman,Gryphon,Dwarvish Fighter,Dwarvish Thunderer,Dwarvish Ulfserker,Dwarvish Scout,Footpad,Fencer,Heavy Infantryman
```
Fencers would look really odd in the Tallin's army of former ragged peasants. I think it's better to assume that the mages kept them to themselves as personal guardsmen.

cooljeanius replied:
I actually initially added the fencers earlier in S06a "Old Friend", when Tallin and friends get the dwarves to forge new arms and armor for them and train them, the reasoning being that swords are newly forged, and that sometimes maps in NR get so crowded that units with skirmisher are necessary. The only reason I'm explicitly listing them in that scenario, too, is because since the recruit list is already explicitly listed there, it should at least be kept consistent with whatever units were previously added to it. If fencers don't seem ragged enough, perhaps some other unit with skirmisher could be created to fill that role instead?

CelticMinstrel replied:
@kabachuha That argument applies equally to heavy infantry as well.

cooljeanius replied:
> @kabachuha That argument applies equally to heavy infantry as well.

My same reply about the fencers applies to heavy infantry as well, except for the part about having skirmisher

kabachuha replied:
> If fencers don't seem ragged enough, perhaps some other unit with skirmisher could be created to fill that role instead?

What about the outlaw rogues? They have the skirmish ability and we already have thugs and footpads on the list, so I think they fit the band.

cooljeanius replied:
> What about the outlaw rogues? They have the skirmish ability and we already have thugs and footpads on the list, so I think they fit the band.

Rogues are level 2, though, and it would seem odd to have those as the one "raw" level 2 unit available for direct recruitment...

kabachuha replied:
Well, if players could recruit Outlaw thiefs, they would be able to advance them into Rogue skirmishers. I don't think it takes too much experience.

cooljeanius replied:
I dunno, I don't really feel like thieves fit, though... maybe earlier in the campaign when they had less options available to them, but at this point in the campaign they have other loyalist units and should be able to forge the high-quality swords necessary for fencers to be able to use

kabachuha replied:
They still have thugs, poachers and footpads for recruit though.

kabachuha replied:
Fencing is not as much about the armor as about the training. I doubt Tallin's men could have mastered fencing in a couple of years the campaign takes place even if they had the best steel.

If they lend them from mages, one or two loyal fencer/HI units may be sufficient and not too breaking story- and gameplaywise.

stevecotton replied:
How about a campaign-specific unit tree? Possibly liminal, using recolored thief sprites (different hair color), and somewhere between a Fencer and an L1 Outlaw?

Among the close-knit community of survivors, thieves had little place. However, a fighting style of nimble footwork suited some more than heavy armor, and no-one would begrudge backstabbing, when said back belonged to a orc.

cooljeanius replied:
> How about a campaign-specific unit tree? Possibly liminal, using recolored thief sprites (different hair color), and somewhere between a Fencer and an L1 Outlaw?
> 
> Among the close-knit community of survivors, thieves had little place. However, a fighting style of nimble footwork suited some more than heavy armor, and no-one would begrudge backstabbing, when said back belonged to a orc.

That might be a good longer-term goal, but it seems a bit out-of-scope for this PR...

cooljeanius replied:
So @Dalas121 recently had a proposal to give skirmisher to the Dwarvish Scout line; if that goes through, I could revert this part without having to invent any new unit...

stevecotton commented on data/campaigns/Northern_Rebirth/scenarios/07a_Settling_Disputes.cfg:
```diff
@@ -158,51 +160,51 @@
         [/modifications]
 
-        hitpoints=1
+        {QUANTITY hitpoints 3 2 1}
```
What's the difference between 1hp and 3hp? Both of them sound like a one-hit kill, and even the hp was raised to taking more than one hit then that would still sound like a design bug - if he's expected to survive any hits at all then there's a reliance on the RNG.

cooljeanius replied:
maybe increase the variance a bit more, then?

stevecotton replied:
I thought the purpose of it was to make the player keep him at the back for this scenario, but it's a long time since I played it.

cooljeanius replied:
> I thought the purpose of it was to make the player keep him at the back for this scenario, but it's a long time since I played it.

Well, he's AI-controlled, and the AI retreats him anyways (has him switching between the keep for recruiting, and the nearest village for healing), so...

cooljeanius replied:
Also, if he manages to survive the scenario, he's added to your recall list and gets auto-recalled in future scenarios, so there's an incentive to try to keep him alive

CelticMinstrel replied:
Ah, if there's an incentive to keep him alive, then this seems fine to me.

cooljeanius replied:
> Ah, if there's an incentive to keep him alive, then this seems fine to me.

should I increase the variance at all, then? Or just keep it as-is?

stevecotton replied:
Well, what's the aim of the change, and does it do that? How would a larger variance change the player's options?

The reason that I asked about this is that it seemed to be just finding some numbers and making them larger, rather than analysing the gameplay and then finding places to tweak how it responds to a lower skill level.

cooljeanius replied:
> Well, what's the aim of the change, and does it do that? How would a larger variance change the player's options?
> 
> The reason that I asked about this is that it seemed to be just finding some numbers and making them larger, rather than analysing the gameplay and then finding places to tweak how it responds to a lower skill level.

The aim of the change is to make it easier to keep Stalrag alive on easier difficulty levels, which is something I had difficulty doing when I played through it. Yes, in some sense it is just finding numbers and making them larger, but I still think doing even that is worth it, though, as it adds replay value when things are different between difficulty levels, as it'll inspire players to try all of them to see what got changed.

stevecotton replied:
Having the aim is good, but the rest of your answer seems to be missing my point. You're jumping from the aim to changing numbers instead of analysing what change to make.

With that aim, maybe the analysis would be "on easy he should always survive one troll getting through his defenses", and then the numbers come from how much a strong troll could do to him in a single turn.

cooljeanius replied:
I think #8337 might be relevant here

FyiurAmron replied:
I agree that 3/2/1 is not a reasonable change since it's just too minor. I'd suggest just going up at least slightly in some way (e.g. healing him after capturing the lich via event or something)

cooljeanius replied:
> With that aim, maybe the analysis would be "on easy he should always survive one troll getting through his defenses", and then the numbers come from how much a strong troll could do to him in a single turn.

Well if I went with a troll, that would be a rather drastic variance... how about if instead we use the weakest unit that's pre-placed for side 3, the Young Ogre. 1 hit from a Young Ogre is 5 damage. So, `{11 6 1}`, then?

knyghtmare commented on data/campaigns/Northern_Rebirth/scenarios/10a_Stolen_Gold.cfg:
```diff
@@ -343,6 +344,14 @@
         [/recall]
 
+        [recall]
+            id=Varem
```
Who is Varem? Player NR like 7 times over the past 10 years and I do not know this character.

cooljeanius replied:
one of the loyal dwarves who joins you when exploring Malifor's dungeons; really it could be any of them, but I just went with him because he's who I had available

knyghtmare replied:
I checked the code and he indeed does appear but the acquisition process is such that most players would not have him unless they prefer battling spiders.

cooljeanius replied:
> If there's multiple loyal dwarves, you could just have it recall one of them at random? The recall tag can take an arbitrary filter.

That's a good idea, actually

cooljeanius replied:
OK, I changed it to recall an arbitrary loyal dwarf in [0149764](https://github.com/wesnoth/wesnoth/pull/6057/commits/014976436016a4f09260a5e12fecc066999ed833) (untested)

cooljeanius commented:
cross-referencing this PR against the feedback I left on the forums: https://forums.wesnoth.org/viewtopic.php?f=38&t=16993&p=660329#p660329

Pentarctagon commented:
@cooljeanius is this still something you're pursuing?

cooljeanius commented:
> @cooljeanius is this still something you're pursuing?

yeah I intend to but it probably won't be this week though

cooljeanius committed:
Update 05a_01_The_Pursuit.map

let's see if this fixes the merge conflict

cooljeanius committed:
Update 05a_01_The_Pursuit.map

remove some stray `^Efs`-es not in mainline

cooljeanius committed:
Update 05a_01_The_Pursuit.map

same as previous

cooljeanius committed:
Update 05a_01_The_Pursuit.map

another one of these

cooljeanius committed:
Update 05a_01_The_Pursuit.map

almost done with this file?

cooljeanius committed:
Update 05a_01_The_Pursuit.map

getting closer

cooljeanius committed:
Merge branch 'master' into egallager_NR_difficulty

cooljeanius commented:
ok I fixed the merge conflict; the failing `map-diff` CI task error message is: 
```
/home/runner/work/wesnoth/wesnoth/utils/wesnoth-map-diff/node_modules/@jimp/utils/dist/index.js:40
  throw error;
  ^

Error: The source must be a Jimp image
    at Jimp.throwError (/home/runner/work/wesnoth/wesnoth/utils/wesnoth-map-diff/node_modules/@jimp/utils/dist/index.js:33:13)
    at Jimp.composite (/home/runner/work/wesnoth/wesnoth/utils/wesnoth-map-diff/node_modules/@jimp/core/dist/composite/index.js:35:30)
    at paintMisc (/home/runner/work/wesnoth/wesnoth/utils/wesnoth-map-diff/build/paint.js:52:16)
    at /home/runner/work/wesnoth/wesnoth/utils/wesnoth-map-diff/build/paint.js:76:17
    at Object.walkthrough (/home/runner/work/wesnoth/wesnoth/utils/wesnoth-map-diff/build/tilemap.js:41:9)
    at paintTilemap (/home/runner/work/wesnoth/wesnoth/utils/wesnoth-map-diff/build/paint.js:73:21)
    at Jimp.<anonymous> (/home/runner/work/wesnoth/wesnoth/utils/wesnoth-map-diff/build/paint.js:84:9)
    at Timeout._onTimeout (/home/runner/work/wesnoth/wesnoth/utils/wesnoth-map-diff/node_modules/@jimp/core/dist/index.js:264:25)
    at listOnTimeout (node:internal/timers:559:17)
    at processTimers (node:internal/timers:502:7)
Error: Process completed with exit code 1.
```
@macabeus do you know anything about that?

macabeus commented:
@cooljeanius I guess it's a bug already fixed on this PR: https://github.com/wesnoth/wesnoth/pull/6880 
If it isn't a fixed bug, we'll at least have a better error message.

cooljeanius committed:
Merge branch 'wesnoth:master' into egallager_NR_difficulty

cooljeanius commented:
> @cooljeanius I guess it's a bug already fixed on this PR: #6880 If it isn't a fixed bug, we'll at least have a better error message.

OK so I updated this pull request to a version where that PR is merged and now the error is:
```
From https://github.com/wesnoth/wesnoth
 * branch              7e32194e43f418b0621af3be297d3c60ea911c7f -> FETCH_HEAD
** Working on data/campaigns/Northern_Rebirth/maps/05a_01_The_Pursuit.map **
/home/runner/work/wesnoth/wesnoth/utils/wesnoth-map-diff/build/images.js:61
                    throw new Error(`Missing image for "^${miscCode}"`);
                          ^

Error: Missing image for "^Xo"
    at Object.getTile (/home/runner/work/wesnoth/wesnoth/utils/wesnoth-map-diff/build/images.js:61:27)
    at paintTile (/home/runner/work/wesnoth/wesnoth/utils/wesnoth-map-diff/build/paint.js:50:34)
    at /home/runner/work/wesnoth/wesnoth/utils/wesnoth-map-diff/build/paint.js:112:17
    at Object.walkthrough (/home/runner/work/wesnoth/wesnoth/utils/wesnoth-map-diff/build/tilemap.js:41:9)
    at paintTilemap (/home/runner/work/wesnoth/wesnoth/utils/wesnoth-map-diff/build/paint.js:110:21)
    at Jimp.<anonymous> (/home/runner/work/wesnoth/wesnoth/utils/wesnoth-map-diff/build/paint.js:123:9)
Error: Process completed with exit code 1.
```

cooljeanius committed:
Merge branch 'wesnoth:master' into egallager_NR_difficulty

cooljeanius committed:
remove and gitignore rejects files

cooljeanius committed:
Merge branch 'egallager_NR_difficulty' of https://github.com/cooljeanius/wesnoth into egallager_NR_difficulty

cooljeanius committed:
Update 01_Breaking_the_Chains.cfg

reword string that uses "reckon" twice to address issue #6510

cooljeanius committed:
Merge branch 'wesnoth:master' into egallager_NR_difficulty

cooljeanius committed:
Merge branch 'wesnoth:master' into egallager_NR_difficulty

cooljeanius committed:
Merge branch 'wesnoth:master' into egallager_NR_difficulty

cooljeanius committed:
Update 01_Breaking_the_Chains.cfg

just tweaking some values; I'm planning on starting another playthru soon, but haven't just yet

stevecotton commented:
> edit: IIRC the main thing getting in the way of me continuing with it was that merge conflict; I wish that wesnoth-map-diff tool could display the parts of maps that are causing a merge conflict, too... I guess I could just discard the changes to that map...

The change that's causing a conflict, 90b68b538872,  is purely a cosmetic change from `Xol` to `Xos` or `Xos^Efs`. So if you've already replaced every terrain that had a pink D, just ignore the changes from 90b68b538872.

cooljeanius commented:
> The change that's causing a conflict, [90b68b5](https://github.com/wesnoth/wesnoth/commit/90b68b5388727ba13cf93bf2fa93f961dcb2e6e3), is purely a cosmetic change from `Xol` to `Xos` or `Xos^Efs`. So if you've already replaced every terrain that had a pink D, just ignore the changes from [90b68b5](https://github.com/wesnoth/wesnoth/commit/90b68b5388727ba13cf93bf2fa93f961dcb2e6e3).

Is there a way to tell the CI to ignore it, too?

stevecotton commented:
Sorry, I thought 870bd715219e319c174ee124af6005c795dc9189 did that - but now I see it failed after than was merged.

A failing build doesn't actually stop the PR being merged.

cooljeanius committed:
Merge branch 'wesnoth:master' into egallager_NR_difficulty

cooljeanius commented on data/campaigns/Northern_Rebirth/scenarios/13a_Showdown.cfg:
```diff
@@ -47,11 +47,13 @@
         team_name=knalgans
         user_team_name=_"Alliance"
-        gold=100
+        {GOLD 900 500 100}
```
@Toranks is this too much?

Toranks replied:
I don't really understand how the gold is distributed in NR yet. I've seen on the forum people with 2000 gold at the start of this scenario on nightmare, I had around 600 with Tallin and it's costing me a lot. I haven't finished the scenario yet.

stevecotton replied:
Did you manage to retrieve the stolen gold that originally came from The Pursuit's treasury?

Toranks replied:
Yes, I retrieve the gold but is calculated and divided in a strange way

stevecotton replied:
Comments in the WML would help, saying why these numbers were chosen.

cooljeanius replied:
> Comments in the WML would help, saying why these numbers were chosen.

I think it was however much was necessary for me to recall my entire recall list on EASY, and then the NORMAL amount was an interpolation of the halfway point between that and the existing HARD amount, but I can't remember for certain...

cooljeanius replied:
Comment added in [743a71c](https://github.com/wesnoth/wesnoth/pull/6057/commits/743a71cb176982fd79d0a2ace9f5c8111f369ad4).

Wedge009 committed:
NR: Give Eryssa hero trait since her death is a loss condition.

Resolves #5374
Resolves #7448

cooljeanius committed:
Merge branch 'wesnoth:master' into egallager_NR_difficulty

cooljeanius committed:
Merge pull request #1 from Wedge009/eryssa_hero

Apply PR wesnoth/wesnoth#7449 to PR wesnoth/wesnoth#6057

TheShadowOfHassen commented:
I don't suppose you could downsize the battles some? I don't mind the difficulty but i'm at the final battle and I have to wait like three minutes for the enemy to move, on fast forward.

cooljeanius commented:
> I don't suppose you could downsize the battles some? I don't mind the difficulty but i'm at the final battle and I have to wait like three minutes for the enemy to move, on fast forward.

Large-scale battles are kind of NR's entire raison d'être; if we removed those, it wouldn't really be NR anymore. I guess there are some AI changes that could make it go faster, though, without actually reducing the scale...

cooljeanius committed:
Seeing if my other PR fixes the map diff failure in this one (#2)

i.e., merge wesnoth#7452 into wesnoth#6057

---------

Co-authored-by: Gunter Labes <soliton@wesnoth.org>
Co-authored-by: Wedge009 <wedge009@wedge009.net>
Co-authored-by: Bruno Macabeus <macabeus@users.noreply.github.com>

cooljeanius committed:
Merge branch 'wesnoth:master' into egallager_NR_difficulty

cooljeanius committed:
Merge https://github.com/wesnoth/wesnoth into egallager_NR_difficulty

cooljeanius committed:
Update map-diff.yml

try adding some additional flags to curl

cooljeanius commented:
> Yay, the map-diff CI task now passes! Looks like it didn't manage to upload the actual image successfully though; that's something that @macabeus and I are going to have to continue to work on... (I've seen it in a few other PRs, too...)

Relevant part of the log:
```
Imgur result:
{"errors":[{"id":"legacy-api-dfc4cf849-jssvj/BYtWSpW8As-1941785","code":"429","status":"Too Many Requests","detail":"Too Many Requests"}]}
```

cooljeanius commented:
blah, there's a merge conflict now... at least it's only in one file so far; if more files start having merge conflicts, I'm just going to give up on this...

Pentarctagon commented:
What remains to be done? To be honest I'm kind of amazed that there's only been a single conflict after a year and a half.

TheShadowOfHassen commented:
Are the conflicts because of my dialog rewriting?

cooljeanius commented:
> What remains to be done? To be honest I'm kind of amazed that there's only been a single conflict after a year and a half.

- Come to some sort of agreement about Fencers and Heavy Infantrymen
- Update recall filter I added for Varem to apply to any loyal dwarves instead
- Another playthrough in which I retest my changes

> Are the conflicts because of my dialog rewriting?

Yeah but don't worry about it; it's ok... I intend to handle it.

FyiurAmron commented on data/campaigns/Northern_Rebirth/scenarios/07a_Settling_Disputes.cfg:
```diff
@@ -125,7 +125,9 @@
-        gold=0 #player has large carryover here, is handled in prestart event
+        # player is supposed to have a large carryover here (handled in prestart event), but just in case:
+        {GOLD 20 10 0}
+        {INCOME 2 1 0}
```
this makes the game harder on NORMAL and much harder on HARD by lowering the income

cooljeanius replied:
`{4 3 2}` then?

cooljeanius replied:
> `{4 3 2}` then?

I went with this in [0149764](https://github.com/wesnoth/wesnoth/pull/6057/commits/014976436016a4f09260a5e12fecc066999ed833).

FyiurAmron commented on data/campaigns/Northern_Rebirth/scenarios/07a_Settling_Disputes.cfg:
```diff
@@ -158,51 +160,51 @@
-            hitpoints=12
-            experience=40
+            {QUANTITY hitpoints 16 14 12}
+            {QUANTITY experience 41 40 39}
```
this makes him weaker on HARD

cooljeanius replied:
I'll have to review this one once I'm back on my computer (on mobile currently)...

cooljeanius replied:
So, right, an Ulfserker's max experience is 42. I can't go any higher without making him auto-advance, and I generally like to keep all 3 numbers different from one another, in a linear pattern...

FyiurAmron replied:
@cooljeanius
> I generally like to keep all 3 numbers different from one another, in a linear pattern

why? There's no need to differentiate the numbers for the sake of just being different, especially if the difference is this minuscule. It makes both maintaining a scenario a chore (more values to change on refactors etc), and, in this and many similar places, serves no actual purpose. I'd mirror what @stevecotton said before: try to find an aim first, and propose the changes for that next.

If the unit is about to get AMLAd anyway, the HPs doesn't matter as long as it's the same amount of troll hits. Same with XP: if the unit is fighting trolls, then 40 XP and 41 XP are effectively the same, they'll either get AMLAd or killed, no other option here.

If you want to make the units able to survive more hits, you'll need to give them more than 2 HPs in this case (AFAIR, troll hits for 10-ish dam). Tweaking XP makes no sense whatsoever, with 40 XP it's just a single fight. That's why you're breaking HARD here, because 39 XP means surviving two troll combats, so it's virtually impossible and you're marking this unit to die.

cooljeanius replied:
> @cooljeanius
> 
> > I generally like to keep all 3 numbers different from one another, in a linear pattern
> 
> why? There's no need to differentiate the numbers for the sake of just being different, especially if the difference is this minuscule.

Maybe not a NEED, but it's still a want, at least.

> It makes both maintaining a scenario a chore (more values to change on refactors etc), and, in this and many similar places, serves no actual purpose.

It increases replayability, though.

> I'd mirror what @stevecotton said before: balancing the scenario is not about having aesthetic numbers. 

I disagree, aesthetic numbers are nice.

> It's about trying to find an aim first, and proposing the changes for that next.
> 
> If the unit is about to get AMLAd anyway, the HPs doesn't matter as long as it's the same amount of troll hits. Same with XP: if the unit is fighting trolls, then 40 XP and 41 XP are effectively the same, they'll either get AMLAd or killed, no other option here.
> 
> If you want to make the units able to survive more hits, you'll need to give them more than 2 HPs in this case (AFAIR, troll hits for 10-ish dam, both 12, 14 and 16 are two hits). Tweaking XP makes no sense whatsoever, with 40 XP it's just a single fight. That's why you're breaking HARD here, because 39 XP means surviving two troll combats, so it's virtually impossible and you're marking this unit to die.

Well anyways, I'm no longer breaking HARD as of [0149764](https://github.com/wesnoth/wesnoth/pull/6057/commits/014976436016a4f09260a5e12fecc066999ed833) (untested).

CelticMinstrel replied:
You do realize that `{QUANTITY 41 40 40}` works just fine, right? Also, does 18 represent more hits than 12?

cooljeanius replied:
> You do realize that `{QUANTITY 41 40 40}` works just fine, right?

Yeah but I don't like repeating numbers between difficulties; I prefer the ifdef method in that case

FyiurAmron commented:
I'm inclined to agree that some changes are really groundbreaking, while some are too minor. I think that doing this per-scenario would be a better way to handle this, this is really huge.

TheShadowOfHassen commented:
So maybe separate the tweaks for each campaign and have a PR for each separately?

I really like the NR campaign, and I'd love to see it get a bit of a balance!

cooljeanius committed:
some changes in response to feedback

stevecotton commented on data/campaigns/Northern_Rebirth/scenarios/07a_Settling_Disputes.cfg:
```diff
@@ -125,7 +125,9 @@
         recruit=Thug,Poacher,Spearman,Bowman,Footpad
-        gold=0 #player has large carryover here, is handled in prestart event
+        # player is supposed to have a large carryover here (handled in prestart event), but just in case:
+        {GOLD 20 10 0}
```
Have you seen a bug here? The code in the prestart seems fine, hence the question about whether there's anything being fixed here.

20 gold seems insignificant - IIRC that "large carryover" is expected to be in the 5000-8000 range.

FyiurAmron replied:
ditto what @stevecotton said: this value is effectively only a placeholder anyway, since prestart event will overwrite it. Differentiating it per difficulty in this place makes no sense.

cooljeanius replied:
I think the change here was to help sniff out if I was in fact seeing a bug with the prestart event or not...

cooljeanius replied:
I put it back to 0 in [d11e74e](https://github.com/wesnoth/wesnoth/pull/6057/commits/d11e74ebe7abe3f47e5391b5b24d31152db65866).

cooljeanius committed:
Update 07a_Settling_Disputes.cfg

put starting gold back to 0

cooljeanius committed:
Merge branch 'master' of https://github.com/wesnoth/wesnoth into egallager_NR_difficulty

cooljeanius committed:
Update copyrights.csv

add line for Xo.png

cooljeanius commented:
> > What remains to be done? To be honest I'm kind of amazed that there's only been a single conflict after a year and a half.
> 
>    * [ ]  Come to some sort of agreement about Fencers and Heavy Infantrymen
> 
>    * [x]  Update recall filter I added for Varem to apply to any loyal dwarves instead (edit: done in [0149764](https://github.com/wesnoth/wesnoth/pull/6057/commits/014976436016a4f09260a5e12fecc066999ed833), but still untested)
> 
>    * [ ]  Another playthrough in which I retest my changes

OK so I fixed the merge conflicts and got the checks passing again; I still want to get checkboxes 1 and 3 resolved before taking this out of draft mode, though...

cooljeanius commented:
Ugh, copyrights got out of sync again...

cooljeanius commented:
> Ugh, copyrights got out of sync again...

@doofus-01 any guidance on how to handle `copyrights.csv` merge conflicts that have arisen due to the change to embed copyright data in the actual images?

doofus-01 commented:
The way I know of is
1.  (optional) `exiftool -g  /path/to/image/image.png` to see what info is in there.  If this is a new image, or was saved with GIMP, there probably isn't anything you want to keep.
2.  `exiftool -all="" /path/to/image/image.png` to clear out cruft metadata added by GIMP or whatever you used
3.  `exiftool -Copyright="GNU GPL v2+" -Artist="unknown;Eric Gallager (cooljeanius);" -CreateDate=now -ModifyDate=now path/to/image/image.png` or whatever is right for the file

2 & 3 can be combined, and there will be an `image.png_original` that you might not need, but that's the cautious way.

There are other ways to edit metadata, but exiftool is what I've been using.  I'm planning to move this stuff to XMP metadata because then it can be read/written in GIMP without the need for this CLI stuff, but it's not there yet.

cooljeanius committed:
Discard changes to copyrights.csv

cooljeanius commented:
> The way I know of is
> 
>     1. (optional) `exiftool -g  /path/to/image/image.png` to see what info is in there.  If this is a new image, or was saved with GIMP, there probably isn't anything you want to keep.

When I do this, it says:
```
Undefined subroutine Image::ExifTool::IsPC called at /opt/sw/lib/perl5/Image/ExifTool.pm line 4982
```

cooljeanius committed:
update metadata with exiftool

cooljeanius commented:
>     3. `exiftool -Copyright="GNU GPL v2+" -Artist="unknown;Eric Gallager (cooljeanius);" -CreateDate=now -ModifyDate=now path/to/image/image.png` or whatever is right for the file

OK I did this, but it doesn't seem to have mollified the CI check...

doofus-01 commented:
PNGs are using XMP now, because that is metadata that GIMP can read/write.  GIMP seems to add extra garbage, so some cleanup is needed if not using Exiftool or similar, but it can be done later.

I think this should work :
```
/opt/sw/bin/exiftool -all="" -XMP:Rights="CC BY-SA 4.0" -XMP:Creator="unknown;Eric Gallager (cooljeanius/egallager);" -XMP:Date=now utils/wesnoth-map-diff/images/Xo.png
```
`-all=""` will remove most of the extra metagarbage, but leave it out if there is some other tags you want to keep.  Use `-XMP:Description="blah"` if there is a comment you want to include.

cooljeanius committed:
Update Xo.png

Update exif info again

cooljeanius commented:
OK, well the CI checks all appear to be passing again now, at least...

cooljeanius commented:
I think instead of leaving this in Draft Mode to indicate that it needs testing, I'm instead going to mark it as "Ready for review" and then apply the "Ready for testing" label...

Dalas121 commented:
As I understand it, the goal here is: _"NR is notoriously difficult and annoying, so this PR attempts to soften it a little bit on the easier difficulties"_

I support that goal, but I'm not sure I agree with the overall direction of this PR.

First, IIRC a lot of NR's difficulty (mostly in the later missions) doesn't come from winning scenarios, but rather from _playing_ scenarios - with massive armies on both sides, just moving units around is really painful. Adding more gold on low difficulties doesn't help with this - I'd prefer to see low-difficulty battle sizes scaled down, rather than up.

Smaller battles on low difficulty would also remove the need for Fencers - I agree with the other commenters that Fencers/HI aren't very thematic, and the NR recruit list is already large. The [Dwarvish Grenadier](https://github.com/wesnoth/wesnoth/issues/9526#issuecomment-2552111562) (if added) will also help with skirmishing.

Second, while I understand that it's nice to have orderly progressions of numbers, I feel many of these changes are so small as to be effectively irrelevant. An extra 1 experience on a unit, a single extra starting village, etc. I won't mandate the remove of these minor changes as I don't feel they're hurting anything (except possibly code clarity), but for the record - I don't see any benefit in having them.

cooljeanius commented:
> As I understand it, the goal here is: _"NR is notoriously difficult and annoying, so this PR attempts to soften it a little bit on the easier difficulties"_
> 
> I support that goal, but I'm not sure I agree with the overall direction of this PR.
> 
> First, IIRC a lot of NR's difficulty (especially in the later missions) doesn't come from winning scenarios, but rather from _playing_ scenarios - with massive armies on both sides, just moving units around is really painful. Adding more gold on low difficulties doesn't help with this - I'd prefer to see low-difficulty (not high-difficulty) battle sizes scaled down, rather than up.
> 

Having massive armies on both sides for huge scale battles is kind of NR's whole reason for existence, though. If you scale the battle sizes down, it wouldn't be NR anymore.

FyiurAmron commented:
> I support that goal, but I'm not sure I agree with the overall direction of this PR.
> 
> First, IIRC a lot of NR's difficulty (especially in the later missions) doesn't come from winning scenarios, but rather from _playing_ scenarios - with massive armies on both sides, just moving units around is really painful. Adding more gold on low difficulties doesn't help with this - I'd prefer to see low-difficulty (not high-difficulty) battle sizes scaled down, rather than up.

this, basically this. NR, apart from tons of bugs and balance problems, is not "difficult" in the strict sense. With a proper combo, you can basically "win" (i.e. pass) many of the scenarios with like, 3-4 units? (leader with the rod and two immortal mages) My 5c: it's difficult because you need to strike those combos and strats precisely, and there's no way to learn them other than brute-forcing the scenarios and triggers etc. (a lot of things come to the player as complete surprise here TBH). 

> Smaller battles on low difficulty would also remove the need for Fencers - I agree with the other commenters that Fencers/HI aren't very thematic, and the NR recruit list is already large. The [Dwarvish Grenadier](https://github.com/wesnoth/wesnoth/issues/9526#issuecomment-2552111562) (if added) will also help with skirmishing.

Again, this. My 5c: Forcingly huge battles with fixed outcome are, frankly, weak. When you realize the outcome can be fixed and that sometimes you don't need to recruit anything or do anything basically but just cheese the scenarios using AI glitches and scenario conditions, it gets tedious instead of fun.

> Second, while I understand that it's nice to have orderly progressions of numbers, I feel many of these changes are so small as to be effectively irrelevant. 1 extra experience on a unit, a village starting captured vs uncaptured, etc. I won't mandate the remove of these minor changes as I don't feel they're hurting anything (except possibly code clarity) - but for the record, I don't see any benefit in having them.

Yet again, this. I've already elaborated on this before, and what I said still holds. These particular changes are not good, they are detrimental; they make managing those scenarios later a bigger chore to other devs, and the "aesthetic" and "replayability" argument for them simply doesn't hold. Also, it breaks consistency with the style of other campaigns.

I still stand by the original suggestion for @cooljeanius - pick the most "difficult" scenario (or the one you care about the most, as you like), and do a MR just for it. It'll be a lot easier for anyone to do an actual review and discussion then. Here, you're mixing apples (difficulty balancing) with oranges (code style and design philosophy changes) at the scope of the _entire mainline campaign_. While I'm not against huge breaking changes personally (actually, I'm _eager_ to see more of those!), it's like they say: _exceptional claims require exceptional proofs_. If you feel like remaking one of the oldest BfW campaigns outside-out, give a better rationale for your changes than "I find those changes aesthetically pleasing" or "people are likely to replay the campaign to see a number increases by one somewhere".

cooljeanius committed:
further scenario updates

- remove the Heavy Infantrymen I had previously added
- additional commentary
- re-linearize enemy incomes in S11

cooljeanius commented:
> > I support that goal, but I'm not sure I agree with the overall direction of this PR.
> > First, IIRC a lot of NR's difficulty (especially in the later missions) doesn't come from winning scenarios, but rather from _playing_ scenarios - with massive armies on both sides, just moving units around is really painful. Adding more gold on low difficulties doesn't help with this - I'd prefer to see low-difficulty (not high-difficulty) battle sizes scaled down, rather than up.
> 
> this, basically this. NR, apart from tons of bugs and balance problems, is not "difficult" in the strict sense. With a proper combo, you can basically "win" (i.e. pass) many of the scenarios with like, 3-4 units? (leader with the rod and two immortal mages) My 5c: it's quite difficult because you need to strike those combos and strats precisely, and there's no way to learn them other than brute-forcing the scenarios and triggers etc. (a lot of things come to the player as complete surprise here TBH).
> 
> > Smaller battles on low difficulty would also remove the need for Fencers - I agree with the other commenters that Fencers/HI aren't very thematic, and the NR recruit list is already large. The [Dwarvish Grenadier](https://github.com/wesnoth/wesnoth/issues/9526#issuecomment-2552111562) (if added) will also help with skirmishing.
> 
> Again, this. My 5c: forcingly huge battles with fixed outcome are, frankly, weak in the sense of gameplay. When you realize the outcome can be fixed and that sometimes you don't need to recruit anything or do anything basically but just cheese the scenarios using AI glitches and scenario conditions, it gets tedious instead of fun.
> 
> > Second, while I understand that it's nice to have orderly progressions of numbers, I feel many of these changes are so small as to be effectively irrelevant. 1 extra experience on a unit, a village starting captured vs uncaptured, etc. I won't mandate the remove of these minor changes as I don't feel they're hurting anything (except possibly code clarity) - but for the record, I don't see any benefit in having them.
> 
> Yet again, this. I've already elaborated on this before, and what I said still holds. These particular changes are not good, they are detrimental; they make managing those scenarios later a bigger chore to other devs, and the "aesthetic" and "replayability" argument for them simply doesn't hold. Also, it breaks consistency with the style of other campaigns.
> 
> I still stand by the original suggestion for @cooljeanius - pick the most "difficult" scenario (or the one you care about the most, as you like), and do a MR just for it. It'll be a lot easier for anyone to do an actual review and discussion then. Here, you're mixing apples (difficulty balancing) with oranges (code style and design philosophy changes) at the scope of the _entire mainline campaign_. While I'm not against huge breaking changes personally (actually, I'm _eager_ to see more of those!), it's like they say: _extraordinary claims require extraordinary evidence_. If you feel like remaking one of the oldest BfW campaigns inside-out, please give a better rationale for your changes than "I find those changes aesthetically pleasing" or "people are likely to replay the campaign to see a number increases by one somewhere". I do really appreciate the amount of effort you put here, but that needs to translate to more than just the effort IMVHO.

In the majority of my previous campaign difficulty PRs (#5268, #5496, #5382, and #5603) I did the entire campaign in one go, and Yumi accepted them like that.

FyiurAmron commented:
> In the majority of my previous campaign difficulty PRs (#5268, #5496, #5382, and #5603) I did the entire campaign in one go, and Yumi accepted them like that.

The scope of changes in those was noticeably smaller (maybe except #5603, but that's understandable since RoW is simply a long campaign), and I those were actually quite reasonable. I wouldn't argue about those ones, honestly. Yet I do that here, simply because I don't want to see NR turned into even worse mess than it is now, and I don't want to stand idly if there's a thing that maybe can be said to affect its future.

Dalas121 commented:
> Having massive armies on both sides for huge scale battles is kind of NR's whole reason for existence, though. If you scale the battle sizes down, it wouldn't be NR anymore.

I agree - but I think there's a little wiggle room on lower difficulties. Both 100 enemies and 80 enemies counts as "huge scale battle" IMO.

For example, on Easy on 05a_01_The_Pursuit.cfg you significantly reduce the bonus gold that enemies gain - I think that's very helpful for a low-skill player. On the other hand, you also increase player gold from 600 to 650 - I think that's mostly irrelevant for a low-skill player, as I don't expect them to have the skill/patience to use 600 gold, much less 650.

I'm not mandating changes here, just elaborating on my campaign balance philosophy and pointing out some changes that I feel are ineffective.

cooljeanius commented:
> For example, on Easy on 05a_01_The_Pursuit.cfg you significantly reduce the bonus gold that enemies gain - I think that's very helpful for a low-skill player. On the other hand, you also increase player gold from 600 to 650 - I think that's mostly irrelevant for a low-skill player, as I don't expect them to have the skill/patience to use 600 gold, much less 650.

That's how much gold I needed to get through the scenario myself; the player is surrounded on all sides by enemies and needs to recall a large number of veterans to survive. Please trust my experiences being a low-skilled player myself when it comes to making adjustments for the benefit of low-skilled players.

Wedge009 commented:
This was marked Postponed in February 2024 and Ready for Testing (Review?) in December 2024. What is the Postponed label for?

Pentarctagon commented:
@Dalas121 any opinion?

Dalas121 commented:
Oh, wow, I'd forgotten about this.

I'm still against inlcuding very minor difficulty variations, like `{QUANTITY hitpoints 32 30 28}` or `{STARTING_VILLAGES 1 {ON_DIFFICULTY 8 7 6}}` - I think those have basically 0 gameplay effect and just serve to complicated the WML. Bigger changes like `{QUANTITY hitpoints 11 6 1}` seem fine to me.

If @cooljeanius is open to removing the minor difficulty variations from this PR, I'm fine merging the rest. I don't think anybody's taken a serious look at NR's balancing in a long time, so it'd be good to update it based on cooljeanius's experiences via this PR. Especially since a lot of unit costs got rebalanced in 1.18.

At some point in the future NR could really benefit from a proper rework, at which point changes made here might become moot. There's no specific timeline for that, though: IIRC there's been numerous UMC attempts and so far no particular one has proved especially popular, so it may still be many years.
